From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tom Sepez <tsepez@google.com>
Date: Sat, 17 Oct 2026 14:38:18 -0700
Subject: [M156] Guard against re-entrant destruction of the CJS_Runtime

Original change's description:
> Guard against re-entrant destruction of the CJS_Runtime
>
> Prevent a (somewhat) misbehaving embedder from tearing down the
> runtime while a callback is still in-flight.
>
> Bug: 572237007
> Change-Id: I003a6b67703305c465a2ed0845f4274df93ed5af
> Reviewed-on: https://pdfium-review.googlesource.com/c/pdfium/+/160614
> Commit-Queue: Lei Zhang <thestig@chromium.org>
> Reviewed-by: Lei Zhang <thestig@chromium.org>

(cherry picked from commit e764cc088a62b5f0510a73bed4e19c88c9af50bc)

Bug: 574749230,572237007
Change-Id: I003a6b67703305c465a2ed0845f4274df93ed5af
Reviewed-on: https://pdfium-review.googlesource.com/c/pdfium/+/161310
Bot-Commit: rubber-stamper@appspot.gserviceaccount.com <rubber-stamper@appspot.gserviceaccount.com>
Auto-Submit: chrome-cherry-picker@chops-service-accounts.iam.gserviceaccount.com <chrome-cherry-picker@chops-service-accounts.iam.gserviceaccount.com>

diff --git a/fxjs/cjs_app.cpp b/fxjs/cjs_app.cpp
index f56b06625cf2a271dfc97a220db654d1b56c2fbc..d18663d595fc96ec28bdf9c6be7e014018781e29 100644
--- a/fxjs/cjs_app.cpp
+++ b/fxjs/cjs_app.cpp
@@ -282,7 +282,10 @@ CJS_Result CJS_App::beep(CJS_Runtime* pRuntime,
     type = pRuntime->ToInt32Reentrant(params[0]);
   }
 
+  pRuntime->BeginBlock();
   pRuntime->GetFormFillEnv()->JS_appBeep(type);
+  pRuntime->EndBlock();
+
   return CJS_Result::Success();
 }
 
@@ -579,8 +582,11 @@ CJS_Result CJS_App::response(CJS_Runtime* pRuntime,
   static constexpr int kMaxWideChars = 1024;
   static constexpr int kMaxBytes = kMaxWideChars * sizeof(uint16_t);
   auto buffer = FixedSizeDataVector<uint8_t>::Zeroed(kMaxBytes);
+
+  pRuntime->BeginBlock();
   int byte_length = pRuntime->GetFormFillEnv()->JS_appResponse(
       swQuestion, swTitle, swDefault, swLabel, bPassword, buffer.span());
+  pRuntime->EndBlock();
 
   if (byte_length < 0 || byte_length > kMaxBytes) {
     return CJS_Result::Failure(JSMessage::kParamTooLongError);
diff --git a/fxjs/cjs_runtime.cpp b/fxjs/cjs_runtime.cpp
index 75c55fa33988a33af3bcc1d3759c23feac6811e7..dc0403ad415924c97dfe183da6ec419abb649a6a 100644
--- a/fxjs/cjs_runtime.cpp
+++ b/fxjs/cjs_runtime.cpp
@@ -86,6 +86,7 @@ CJS_Runtime::CJS_Runtime(CPDFSDK_FormFillEnvironment* pFormFillEnv)
 }
 
 CJS_Runtime::~CJS_Runtime() {
+  CHECK(!IsBlocking());
   NotifyObservers();
   ReleaseEngine();
   if (isolate_managed_) {
